File: Motocitizen/src/motocitizen/ui/popups/PopupWindowGeneral.java
package motocitizen.ui.popups;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.PopupWindow;
import android.widget.TableLayout;
import android.widget.TableRow;

import com.google.android.gms.maps.model.LatLng;

import kotlin.Unit;
import motocitizen.content.accident.Accident;
import motocitizen.datasources.network.ApiResponse;
import motocitizen.datasources.network.requests.ActivateAccident;
import motocitizen.datasources.network.requests.BanRequest;
import motocitizen.datasources.network.requests.EndAccident;
import motocitizen.datasources.network.requests.HideAccident;
import motocitizen.main.R;
import motocitizen.router.Router;
import motocitizen.utils.ToastUtils;

abstract public class PopupWindowGeneral extends PopupWindow {
    private static final ColorDrawable         windowBackGround = new ColorDrawable(0x00ffffff);
    private static final int                   contentColor     = 0xFF202020;
    final protected      TableRow.LayoutParams layoutParams     = new TableRow.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
    private         Context     context;
    final protected TableLayout rootView;

    PopupWindowGeneral(Context context) {
        this.context = context;
        rootView = new TableLayout(context);
        rootView.setOrientation(LinearLayout.HORIZONTAL);
        rootView.setBackgroundColor(contentColor);
        rootView.setLayoutParams(layoutParams);
        setContentView(rootView);
        setWidth(LayoutParams.WRAP_CONTENT);
        setHeight(LayoutParams.WRAP_CONTENT);
        setBackgroundDrawable(windowBackGround);
        setOutsideTouchable(true);
    }

    protected TableRow shareMessageView(final String textToShare) {
        return makeButton(R.string.share, v -> shareButtonPressed(textToShare));
    }

    private void shareButtonPressed(String text) {
        Router.INSTANCE.share((Activity) context, text);
        dismiss();
    }

    protected TableRow copyButtonView(String text) {
        return makeButton(R.string.copy, v -> copyButtonPressed(text));
    }

    private void copyButtonPressed(String text) {
        ClipboardManager myClipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        myClipboard.setPrimaryClip(ClipData.newPlainText("text", text));
        dismiss();
    }

    protected TableRow phoneButtonView(String phone) {
        return makeButton(context.getString(R.string.popup_dial, phone), v -> phoneButtonPressed(phone));
    }

    private void phoneButtonPressed(String phone) {
        Router.INSTANCE.dial((Activity) context, phone);
        dismiss();
    }

    protected TableRow smsButtonView(String phone) {
        return makeButton(context.getString(R.string.popup_sms, phone), v -> smsButtonPressed(phone));
    }

    private void smsButtonPressed(String phone) {
        Router.INSTANCE.sms((Activity) context, phone);
        dismiss();
    }

    protected TableRow finishButtonView(Accident point) {
        return makeButton(point.isEnded() ? R.string.unfinish : R.string.finish, v -> finishButtonPressed(point));
    }

    private void finishButtonPressed(Accident point) {
        if (point.isEnded()) {
            new ActivateAccident(point.getId(), response -> Unit.INSTANCE);
        } else {
            new EndAccident(point.getId(), response -> Unit.INSTANCE);
        }
        dismiss();
    }

    protected TableRow hideButtonView(Accident point) {
        return makeButton(point.isHidden() ? R.string.show : R.string.hide, v -> hideButtonPressed(point));
    }

    private void hideButtonPressed(Accident point) {
        if (point.isHidden()) {
            new ActivateAccident(point.getId(), response -> Unit.INSTANCE);
        } else {
            new HideAccident(point.getId(), response -> Unit.INSTANCE);
        }
        dismiss();
    }

    protected TableRow coordinatesButtonView(Accident point) {
        return makeButton(R.string.copy_coordinates, v -> coordinatesButtonPressed(point));
    }

    private void coordinatesButtonPressed(Accident point) {
        LatLng           latlng      = point.getCoordinates();
        ClipboardManager myClipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        myClipboard.setPrimaryClip(ClipData.newPlainText("text", String.format("%s,%s", latlng.latitude, latlng.longitude)));
        ToastUtils.show(context, context.getString(R.string.coordinates_copied));
        dismiss();
    }

    protected TableRow banButtonView(int id) {
        return makeButton("Забанить", v -> banButtonPressed(id));
    }

    private void banButtonPressed(int id) {
        new BanRequest(id, this::banRequestCallback);
        dismiss();
    }

    private Unit banRequestCallback(ApiResponse response) {
        ((Activity) context).runOnUiThread(
                () -> ToastUtils.show(context,
                                      response.hasError()
                                      ? "Ошибка связи с сервером"
                                      : "Пользователь забанен")
                                          );
        return Unit.INSTANCE;
    }

    private TableRow makeButton(String text, View.OnClickListener listener) {
        Button button = new Button(context);
        button.setText(text);
        button.setOnClickListener(listener);
        TableRow tableRow = new TableRow(context);
        tableRow.addView(button, layoutParams);
        return tableRow;
    }

    private TableRow makeButton(int text, View.OnClickListener listener) {
        return makeButton(context.getString(text), listener);
    }
}

File: Motocitizen/src/motocitizen/ui/popups/PopupWindowGeneral.kt
package motocitizen.ui.popups

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.view.View
import android.view.ViewGroup.LayoutParams
import android.widget.*
import motocitizen.content.accident.Accident
import motocitizen.datasources.network.ApiResponse
import motocitizen.datasources.network.requests.ActivateAccident
import motocitizen.datasources.network.requests.BanRequest
import motocitizen.datasources.network.requests.EndAccident
import motocitizen.datasources.network.requests.HideAccident
import motocitizen.main.R
import motocitizen.router.Router
import motocitizen.utils.isEnded
import motocitizen.utils.isHidden
import motocitizen.utils.show

abstract class PopupWindowGeneral internal constructor(private val context: Context) : PopupWindow() {
    private val windowBackGround = ColorDrawable(0x00ffffff)
    private val contentColor = 0xFF202020.toInt()

    protected val layoutParams = TableRow.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
    protected val rootView: TableLayout = TableLayout(context)

    init {
        rootView.orientation = LinearLayout.HORIZONTAL
        rootView.setBackgroundColor(contentColor)
        rootView.layoutParams = layoutParams
        contentView = rootView
        width = LayoutParams.WRAP_CONTENT
        height = LayoutParams.WRAP_CONTENT
        setBackgroundDrawable(windowBackGround)
        isOutsideTouchable = true
    }

    protected fun shareMessageView(textToShare: String) = makeButton(R.string.share, { shareButtonPressed(textToShare) })

    private fun shareButtonPressed(text: String) {
        Router.share(context as Activity, text)
        dismiss()
    }

    protected fun copyButtonView(text: String) = makeButton(R.string.copy, { copyButtonPressed(text) })

    private fun copyButtonPressed(text: String) {
        val myClipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        myClipboard.primaryClip = ClipData.newPlainText("text", text)
        dismiss()
    }

    protected fun phoneButtonView(phone: String) = makeButton(context.getString(R.string.popup_dial, phone), { phoneButtonPressed(phone) })

    private fun phoneButtonPressed(phone: String) {
        Router.dial(context as Activity, phone)
        dismiss()
    }

    protected fun smsButtonView(phone: String) = makeButton(context.getString(R.string.popup_sms, phone), { smsButtonPressed(phone) })

    private fun smsButtonPressed(phone: String) {
        Router.sms(context as Activity, phone)
        dismiss()
    }

    protected fun finishButtonView(point: Accident) = makeButton(if (point.isEnded()) R.string.unfinish else R.string.finish, { finishButtonPressed(point) })

    private fun finishButtonPressed(point: Accident) {
        if (point.isEnded()) {
            ActivateAccident(point.id) { }
        } else {
            EndAccident(point.id) { }
        }
        dismiss()
    }

    protected fun hideButtonView(point: Accident) = makeButton(if (point.isHidden()) R.string.show else R.string.hide, { hideButtonPressed(point) })

    private fun hideButtonPressed(point: Accident) {
        if (point.isHidden()) {
            ActivateAccident(point.id) { }
        } else {
            HideAccident(point.id) { }
        }
        dismiss()
    }

    protected fun coordinatesButtonView(point: Accident): TableRow = makeButton(R.string.copy_coordinates, { coordinatesButtonPressed(point) })

    private fun coordinatesButtonPressed(point: Accident) {
        val latLng = point.coordinates
        val myClipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        myClipboard.primaryClip = ClipData.newPlainText("text", String.format("%s,%s", latLng.latitude, latLng.longitude))
        show(context, context.getString(R.string.coordinates_copied))
        dismiss()
    }

    protected fun banButtonView(id: Int): TableRow = makeButton("Забанить", { banButtonPressed(id) })

    private fun banButtonPressed(id: Int) {
        BanRequest(id, this::banRequestCallback)
        dismiss()
    }

    private fun banRequestCallback(response: ApiResponse) {
        (context as Activity).runOnUiThread {
            show(context, if (response.hasError())
                "Ошибка связи с сервером"
            else
                "Пользователь забанен")
        }
    }

    private fun makeButton(text: String, listener: (View) -> Unit): TableRow {
        val button = Button(context)
        button.text = text
        button.setOnClickListener(listener)
        val tableRow = TableRow(context)
        tableRow.addView(button, layoutParams)
        return tableRow
    }

    private fun makeButton(text: Int, listener: (View) -> Unit): TableRow = makeButton(context.getString(text), listener)
}
